### PR TITLE
Refactor IndieAuth function, add type hints, modify documentation to mention new type classes

### PR DESCRIPTION
**Pull request by capjamesg**
This PR refactors the IndieAuth function to use a new set of exceptions and consistent return values. Type hints have been added to many functions where they were not previously used, with some code cleaning done to ensure long lines were not built up for the longer functions.

I have also rewritten the documentation for some functions to make use of Sphinx' autoclass generator. This will let us automatically show return objects. This change replaces the documentation for JSON objects that have are being made obsolete during the migration to objects.

### CODE REVIEW

File: src/indieweb_utils/indieauth/flask.py
Comment (jamesvandyne) [inline src/indieweb_utils/indieauth/flask.py]: The proper typing for a dictionary is to use the `Dict` class from the `typing` module. That also allows us to type the key / value. I'm not sure what the expected response is, so maybe we need to use `Any` for the value?

```python
from typing import Dict, Any
...
endpoint_response: Dict[str, Any]
```

File: src/indieweb_utils/indieauth/flask.py
Comment (jamesvandyne) [inline src/indieweb_utils/indieauth/flask.py]: The docstrings indicate that required is a List of strings. 

```python
from typing import List

required_scopes: List[str]
```

What do you think about making this function kwarg only? i.e. It must be called like

```python
indieauth_callback_handler(code="hogehoge", state="fuga"...)
```

If you think that's a good idea, make the first argument a *

```python
def indieauth_callback_handler(
    *,
    code: str,
    ...
```

File: src/indieweb_utils/indieauth/flask.py
Comment (jamesvandyne) [inline src/indieweb_utils/indieauth/flask.py]: We could probably split this into a separate function that could be re-used for validating the response.

File: src/indieweb_utils/indieauth/flask.py
Comment (jamesvandyne) [inline src/indieweb_utils/indieauth/flask.py:74]: I wonder if the "making a request" part of this couldn't be broken into a smaller function that could be re-used.

File: src/indieweb_utils/posts/discovery.py
Comment (jamesvandyne) [inline src/indieweb_utils/posts/discovery.py:104]: For a different PR, but I wonder if we couldn't make this return our Author object?

File: src/indieweb_utils/webmentions/send.py
Comment (jamesvandyne) [inline src/indieweb_utils/webmentions/send.py]: Other attributes (title, description, url) of SendWebmentionResponse are nouns, so we should probably rename succeeded -> success to keep it consistent.
```suggestion
            success=False,
```

File: docs/source/usage.rst
Comment (jamesvandyne) [inline docs/source/usage.rst:208]: ❤️ Love it. Thank you!

File: src/indieweb_utils/posts/discovery.py
Comment (capjamesg) [reply src/indieweb_utils/posts/discovery.py:104]: The h-card microformat has a lot of potential values, including one that is nested: https://microformats.org/wiki/h-card

I think typeclasses are easier to read but it's worth discussing this since we will have to introduce quite a bit of repetitive code in the return function to accomodate for all the values in a h-card.

Comment (jamesvandyne) [reply src/indieweb_utils/posts/discovery.py:104]: We shouldn't end up with too much redundant code. My thinking is that

1. We should have h-card properties listed in a constants file (just the common 4 or 5 properties is fine to start). We'd also have a list of "valid properties".
2. Loop through the passed properties and store the valid ones in a dict
3. Use that dictionary to populate the dataclass

```python
import dataclasses

@dataclasses.dataclass
class Author:
    ...
    p_job_title: str = dataclasses.field(default="")
    # Or maybe it Optional[str] and set default to None?


def discover_author(..):
    author_properties ={key: value for key, value in properties.values() if key in valid_author_properties} 
    ...
    # pass the dict as kwargs to the class
    return Author(**author_properties)  
```

I'll try and get a PR together for it once this merges.

Comment (capjamesg) [reply src/indieweb_utils/posts/discovery.py:104]: Thank you @jamesvandyne! I pushed all of your feedback in #e006147. Pending your approval, we can merge this PR into the main branch so we can get to work on the h-card / authorship representation change.

**Review by jamesvandyne — APPROVED** [review]
LGTM!

**Pull request merged**
